lib/n4l/consul_balancer.lua

```diff
@@ -196,35 +202,17 @@ function _M.round_robin(service_name)
     ngx.log(ngx.ERR, "consul.balancer: no entry found for service: ", service_name)
     return ngx.exit(500)
   end
-  if service.upstreams == nil or #service.upstreams == 0 then
-    ngx.log(ngx.ERR, "consul.balancer: no peers for service: ", service_name)
+  local rr_up = service.rr_up
+  if rr_up == nil then
+    ngx.log(ngx.ERR, "consul.balancer: no roundrobin object for service: ", service_name)
     return ngx.exit(500)
   end
-  if service.state == nil or service.state > #service.upstreams then
-    service.state = 1
-  end
-  -- TODO: https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#get_last_failure
-  -- set max tries only at first attempt
-  if not balancer.get_last_failure() then
-    balancer.set_more_tries(#service.upstreams - 1)
-  end
-  -- Picking next upstream
-  local upstream = service.upstreams[service.state]
-  service.state = service.state + 1
-  _persist(service_name, service)
-  local ok, err = balancer.set_current_peer(upstream["address"], upstream["port"])
+  local server = rr_up:find()
+  local ok, err = balancer.set_current_peer(server)
```

Comment:
I could not find firm proof that this function behaves correctly if passed a string containing ':' separated host/port pair.

I checked here: https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.md#set_current_peer
and here: https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/balancer.lua#L109

Please verify this by testing.

Reply:
The `find()` method does return a string of `ip:port`, and as you can see from the [lua-resty-balancer](https://github.com/openresty/lua-resty-balancer) example, the `ip:port` string can be passed directly to the `set_current_peer()` method as an argument.

After testing, it is feasible, it is possible that the `set_current_peer()` method is internally compatible.